### PR TITLE
git.get_commits error check

### LINKED ISSUE #518: `cz changelog` is not reporting git errors (`git.py` missing `cmd.return_code` checks)
### Description

`cz commit` worked just fine, but `cz changelog` resulted in the error:
> "No commits found"

I later came to find the calls to git were silently failing in `git.get_tags` due to an unsupported version of git.
It's not very obvious the issue is with git.
Might we want some kind of warnings/exceptions [here](https://github.com/commitizen-tools/commitizen/blob/afa0d93544405bb0a871b5072e0a814720d933be/commitizen/git.py#L123)"?

### Steps to reproduce

1. unsupported git (e.g. 1.8.3.1) (normally we use a newer version)
2. `cz changelog` (`cz commit` worked)

### Current behavior

Error handling could be improved.

### Desired behavior

Maybe a warning or `raise` would help users out when `cmd.run( ...git...)` fails?

### Screenshots

_No response_

### Environment

CentOS7

### PR DESCRIPTION
**Pull request by bhelgs**
## Description
`git.get_commits` could potentially fail silently. 

The first commit is the fix. 
See the git commit message for more info on all following commits (broken up for easier review). 
image

The final commit also improves error checking in `git.get_tags`  (closes #518).

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

## Expected behavior
Fail fast behavior and better reports to the user when something goes wrong with git.

Some of the tests showed the command was failing for unexpected reasons (e.g. invalid commit range `git log None..None`).

#### Example of a silent failure that is now caught:
Without this change users with really old versions of git will see is "No commits found" but after the change: <b>
`GitCommandError: fatal: unrecognized argument: --author-date-order`
* reason:  `commands/changelog.py` calls `git.get_commits` with `--author-date-order`  but this argument is not valid for older versions of it. 

## Steps to Test This Pull Request
1. ran tests locally and formatted.

### CODE REVIEW

File: docs/exit_codes.md
Comment (bhelgs) [inline docs/exit_codes.md:8]: I wonder why my formatter did this. Not sure if this is better.

File: tests/commands/test_changelog_command.py
Comment (Lee-W) [inline tests/commands/test_changelog_command.py:29]: If we need to compare the error message of external tool (e.g., `git`), we'll need to take `LC_ALL` into consideration. Otherwise, people with difference `LC_ALL` will not pass the test

Comment (Lee-W) [reply tests/commands/test_changelog_command.py:29]: We could try something like changing the git commit command to `LC_ALL=en_US.UTF-8 git commit`

Comment (bhelgs) [reply tests/commands/test_changelog_command.py:29]: you bring up a really good point. Are you happy as-is? How would you like to proceed with this PR?

Comment (Lee-W) [reply tests/commands/test_changelog_command.py:29]: I notice you join the discussion [here](https://github.com/commitizen-tools/commitizen/pull/543#issuecomment-1203038161). Maybe we could discuss this kind of issue there 🤔 

> How would you like to proceed with this PR?

Yep, I'm ok with asserting the message. 

> How would you like to proceed with this PR?

We'll need a way to resolve the language issue. After that, I think we're almost ready to merge

Comment (bhelgs) [reply tests/commands/test_changelog_command.py]: What about this approach of retrying the command?

1. `execinfo` captures the command that failed.
2. test retries executing the command with `LC_ALL`.
3. assert the retry matches the expected `en_US` output of the command.

``` python
    with pytest.raises(GitCommandError) as execinfo:
         cli.main()

     c = cmd.run("LC_ALL=en_US.UTF-8 " + execinfo.value.cmd)
     assert "does not have any commits yet" in c.err
```

File: tests/commands/test_changelog_command.py
Comment (Lee-W) [inline tests/commands/test_changelog_command.py:691]: Do we need to use the approach above as well?

File: tests/commands/test_changelog_command.py
Comment (Lee-W) [reply tests/commands/test_changelog_command.py]: I'm good with this approach :) 

@Woile what do you think. Different from what's discussed on #543  `LC_ALL` part only happens in test cases.

File: tests/commands/test_changelog_command.py
Comment (bhelgs) [reply tests/commands/test_changelog_command.py:691]: In this case `commitizen/changelog.py` is what raised the exception message (in English) so no need to have `LC_ALL`:

image

File: tests/commands/test_changelog_command.py
Comment (bhelgs) [reply tests/commands/test_changelog_command.py]: I wonder if I should have just added `LC_ALL=en_US.UTF-8` to the entire testing environment.

Comment (bhelgs) [reply tests/commands/test_changelog_command.py]: that would be quite a bit more straightforward than this retry strategy.

Comment (bhelgs) [reply tests/commands/test_changelog_command.py]: assuming we don't mock out that specific git call entirely.

Comment (Lee-W) [reply tests/commands/test_changelog_command.py]: If it's just for testing, we could use [pytest-env](https://pypi.org/project/pytest-env/) to achieve it. If it's for the functional code, maybe we'll need to think of a better solution like we did on #543

File: tests/commands/test_changelog_command.py
Comment (Lee-W) [reply tests/commands/test_changelog_command.py:691]: ah got it. Thanks @bhelgs !

File: tests/commands/test_changelog_command.py
Comment (bhelgs) [reply tests/commands/test_changelog_command.py]: @Lee-W After much thought, this test probably isn't providing much on the `command` level.
I think all we want to be concerned about is that we handle an empty git log properly. 

I have refactored `get_commits` so that it calls a `get_log` function that we can test as it's own unit. https://github.com/commitizen-tools/commitizen/pull/540/commits/4317d9f3aba9cb5edba702f0076155118578fad1

Comment (bhelgs) [reply tests/commands/test_changelog_command.py]: As in:

**tests/test_git.py**
``` python
@pytest.mark.usefixtures("tmp_commitizen_project")
 def test_get_log_as_str_list_empty():
     """ensure an exception or empty list in an empty project"""
     try:
         gitlog = git._get_log_as_str_list(start=None, end="HEAD", args="")
     except exceptions.GitCommandError:
         return
     assert len(gitlog) == 0, "list should be empty if no assert"
```

replaces: <br>

**tests/commands/test_changelog_command.py**
``` python
@pytest.mark.usefixtures("tmp_commitizen_project")
 def test_changelog_on_empty_project(mocker):
     testargs = ["cz", "changelog", "--dry-run"]
     mocker.patch.object(sys, "argv", testargs)

     with pytest.raises(GitCommandError):
         cli.main()

     #  git will error out with something like:
     #  "your current branch 'XYZ' does not have any commits yet"
     #  is it wise to assert the exception contains a message like this?
```

Comment (bhelgs) [reply tests/commands/test_changelog_command.py]: And then no need for the `LC_ALL` at all since the function being tested is so small in comparison to before when we tested the whole command

File: tests/commands/test_changelog_command.py
Comment (Lee-W) [inline tests/commands/test_changelog_command.py:18]: May i know why is this test removed?

**Review by Lee-W — COMMENTED** [review]
Thanks @bhelgs ! I'm good with most of the stuff. Just leave one question and we might need your help to rebase from the master branch. I think we're almost ready to merge 💪

File: tests/commands/test_changelog_command.py
Comment (bhelgs) [reply tests/commands/test_changelog_command.py:18]: This test would be a duplicate of `test_git.py::test_get_log_as_str_list_empty`. 
Also, it seems unnecessary for the changelog command/test to enforce a specific exception since it has no way of handling such an exception.

File: commitizen/git.py
Comment (bhelgs) [inline commitizen/git.py:125]: I wonder if this `\n` was working properly for Windows.

File: tests/commands/test_changelog_command.py
Comment (Lee-W) [reply tests/commands/test_changelog_command.py:17]: ah got it. Thanks for your explanation!

**Review by Lee-W — APPROVED** [review]
@Woile I'm good with this one as well. Could you please take a look when you have time? Thanks!

File: commitizen/git.py
Comment (Lee-W) [reply commitizen/git.py:125]: Our CI runs on Windows machine as well so I think it should be fine 🤔 But both maintainers do not have windows machine to test it out. Might need others' help to verify it.

**Comment by Woile** [comment]
Thanks a lot!

**Pull request merged**
